java/server/src/org/openqa/selenium/grid/sessionqueue/GetNewSessionResponse.java

```diff
@@ -0,0 +1,135 @@
+    });
+
+    this.bus.addListener(NEW_SESSION_REJECTED, event -> {
+      try {
+        NewSessionErrorResponse sessionResponse = event.getData(NewSessionErrorResponse.class);
+        this.setErrorResponse(sessionResponse);
+      } catch (Exception ignore) {
+        // Ignore any exception. Do not want to block the eventbus thread.
+      }
+    });
+  }
+
+  private void setResponse(NewSessionResponse sessionResponse) {
+    // Each thread will get its own CountDownLatch and it is stored in the Map using request id as the key.
+    // EventBus thread will retrieve the same request and set it's response and unblock waiting request thread.
+    UUID id = sessionResponse.getRequestId();
+    NewSessionRequest sessionRequest = knownRequests.get(id);
+    sessionRequest.setSessionResponse(
+        new HttpResponse().setContent(bytes(sessionResponse.getDownstreamEncodedResponse())));
+    sessionRequest.getLatch().countDown();
+  }
+
+  private void setErrorResponse(NewSessionErrorResponse sessionResponse) {
+    UUID id = sessionResponse.getRequestId();
+    NewSessionRequest sessionRequest = knownRequests.get(id);
+    sessionRequest
+        .setSessionResponse(new HttpResponse()
+                                .setStatus(HTTP_INTERNAL_ERROR)
+                                .setContent(asJson(
+                                    ImmutableMap.of("message", sessionResponse.getMessage()))));
+    sessionRequest.getLatch().countDown();
+  }
+
+  public HttpResponse add(HttpRequest request) {
+    Require.nonNull("New Session request", request);
+
+    CountDownLatch latch = new CountDownLatch(1);
+    UUID requestId = UUID.randomUUID();
+    NewSessionRequest requestIdentifier = new NewSessionRequest(requestId, latch);
+    knownRequests.put(requestId, requestIdentifier);
+
+    if (!sessionRequests.offerLast(request, requestId)) {
+      return new HttpResponse()
+          .setStatus(HTTP_INTERNAL_ERROR)
+          .setContent(asJson(ImmutableMap.of("message",
+                                             "Session request could not be created. Error while adding to the session queue.")));
+    }
+
+    try {
+      // Block until response is received.
+      // This will not wait indefinitely due to request timeout handled by the LocalDistributor.
+      latch.await();
+      HttpResponse res = requestIdentifier.getSessionResponse();
+      removeRequest(requestId);
+      return res;
+    } catch (InterruptedException e) {
+      LOG.log(Level.WARNING, "The thread waiting for new session response interrupted. {0}",
+              e.getMessage());
+      Thread.currentThread().interrupt();
+
+      return new HttpResponse()
+          .setStatus(HTTP_INTERNAL_ERROR)
+          .setContent(asJson(ImmutableMap.of("message",
+                                             "Session request could not be created. Error while processing the session request.")));
+    }
+}
+
+  private void removeRequest(UUID id) {
+    knownRequests.remove(id);
```

Comment:
What happens if we're in the middle of waiting for `add` to finish?

Reply:
Good point. We shouldn't block the response. Will fix it.